subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +586,476 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
+		return true;
 	}
 
-	LOG_DBG("last frag");
-	tx_data(buf)->is_cont = true;
-	return send_frag(conn, buf, NULL, FRAG_END);
+	return false;
 }
 
-static struct k_poll_signal conn_change =
-		K_POLL_SIGNAL_INITIALIZER(conn_change);
-
-static void conn_cleanup(struct bt_conn *conn)
+bool bt_data_should_stop(struct bt_conn *conn)
 {
-	struct net_buf *buf;
-
-	/* Give back any allocated buffers */
-	while ((buf = net_buf_get(&conn->tx_queue, K_NO_WAIT))) {
-		struct bt_conn_tx *tx = tx_data(buf)->tx;
+	LOG_DBG("%p", conn);
 
-		tx_data(buf)->tx = NULL;
-
-		/* destroy the buffer */
-		net_buf_unref(buf);
-
-		/* destroy the tx context (and any associated meta-data) */
-		if (tx) {
-			conn_tx_destroy(conn, tx);
-		}
+	/* TODO: This function should be overridable by the application: they
+	 * should be able to provide their own heuristic.
+	 */
+	if (!conn_has_data(conn)) {
+		LOG_DBG("no more data");
+		return true;
 	}
 
-	__ASSERT(sys_slist_is_empty(&conn->tx_pending), "Pending TX packets");
-	__ASSERT_NO_MSG(conn->pending_no_cb == 0);
+	/* Queue only 3 buffers per-conn for now */
+	if (atomic_get(&conn->in_ll) < 3) {
+		/* The goal of this heuristic is to allow the link-layer to
+		 * extend an ACL connection event as long as the application
+		 * layer can provide data.
+		 *
+		 * Here we chose three buffers, as some LLs need two enqueued
+		 * packets to be able to set the more-data bit, and one more
+		 * buffer to allow refilling by the app while one of them is
+		 * being sent over-the-air.
+		 */
+		LOG_DBG("should queue more");
+		return false;
+	}
 
-	bt_conn_reset_rx_state(conn);
+	LOG_DBG("conn should stop");
 
-	k_work_reschedule(&conn->deferred_work, K_NO_WAIT);
+	return true;
 }
 
-static void conn_destroy(struct bt_conn *conn, void *data)
+void bt_conn_data_ready(struct bt_conn *conn)
 {
-	if (conn->state == BT_CONN_CONNECTED ||
-	    conn->state == BT_CONN_DISCONNECTING) {
-		bt_conn_set_state(conn, BT_CONN_DISCONNECT_COMPLETE);
+	LOG_DBG("DR");
+
+	/* The TX processor will call the `pull_cb` to get the buf */
+	if (!atomic_set(&conn->_conn_ready_lock, 1)) {
+		sys_slist_append(&bt_dev.le.conn_ready,
+				 &conn->_conn_ready);
+		LOG_DBG("raised");
+	} else {
+		LOG_DBG("already in list");
 	}
 
-	bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+	/* Kick the TX processor */
+	bt_tx_irq_raise();
 }
 
-void bt_conn_cleanup_all(void)
+static bool cannot_send_to_controller(struct bt_conn *conn)
 {
-	bt_conn_foreach(BT_CONN_TYPE_ALL, conn_destroy, NULL);
+	return k_sem_count_get(bt_conn_get_pkts(conn)) == 0;
 }
 
-static int conn_prepare_events(struct bt_conn *conn,
-			       struct k_poll_event *events)
+struct bt_conn *get_conn_ready(void)
 {
-	if (!atomic_get(&conn->ref)) {
-		return -ENOTCONN;
+	/* Here we only peek: we pop the conn (and insert it at the back if it
+	 * still has data) after the QoS function returns false.
+	 */
+	sys_snode_t *node  = sys_slist_peek_head(&bt_dev.le.conn_ready);
```

Comment:
how about this: To avoid getting derailed with fighting the (now very slow) CI in this PR on this, I make a note and submit a follow-up PR constifying this diff? How does that sound?